common/src/main/java/io/netty/util/concurrent/CompleteFuture.java

```diff
@@ -51,20 +51,24 @@ public EventExecutor executor() {
     @Override
     public Future<V> addListener(GenericFutureListener<? extends Future<? super V>> listener) {
         requireNonNull(listener, "listener");
-        DefaultPromise.notifyListener(executor(), this, listener);
+        DefaultPromise.safeExecute(executor(), () -> DefaultPromise.notifyListener0(this, listener));
         return this;
     }
 
     @Override
     public Future<V> addListeners(GenericFutureListener<? extends Future<? super V>>... listeners) {
         requireNonNull(listeners, "listeners");
-        for (GenericFutureListener<? extends Future<? super V>> l: listeners) {
+        DefaultPromise.safeExecute(executor(), () -> notifyListeners(listeners));
+        return this;
+    }
+
+    private void notifyListeners(GenericFutureListener<? extends Future<? super V>>... listeners) {
+        for (GenericFutureListener<? extends Future<? super V>> l : listeners) {
```

Comment:
Can we assert the thread?

Reply:
@johnou you mean that it is `inEventLoop() == true` ?

Reply:
Yep, I think it's quite common through the Netty codebase?